### PR TITLE
rbd: implement new 'rbd perf image iostat/iotop' commands

### PR DESCRIPTION
**Pull request by dillaman**
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by dillaman** [comment]
jenkins test docs

File: src/tools/rbd/ArgumentTypes.cc [hunk 1]
```diff
@@ -166,20 +166,19 @@ void add_snap_id_option(po::options_description *opt) {
...
-}
-
-void add_namespace_options(boost::program_options::options_description *pos,
-                           boost::program_options::options_description *opt) {
-  if (pos != nullptr) {
+  if (namespaces_supported) {
+    add_namespace_option(opt, ARGUMENT_MODIFIER_NONE);
+    pos->add_options()
+      ("pool-spec", "pool specification\n"
+       "(example: <pool-name>/[<namespace-name>]");
```
[the remark this answers is not in the document]
Comment (trociny) [reply src/tools/rbd/ArgumentTypes.cc]: Ah already fixed

**Comment by dillaman** [comment]
> @dillaman It looks like iotop header is not aligned very well with value. I observe this:

I have the titles right-aligned and the numeric values left-aligned (like the ```iostat``` table output). I have it fixed to a 12 character width for each metric column.

**Comment by trociny** [comment]
> > @dillaman It looks like iotop header is not aligned very well with value. I observe this:
> 
> I have the titles right-aligned and the numeric values left-aligned (like the `iostat` table output). I have it fixed to a 12 character width for each metric column.

Why not to align them the same way (both titles and values for counters left aligned and for the image name left aligned)? The output looks more readable for me then.

And for `rbd perf image iostat` output, here is example:
```
NAME        WR    RD    WR_BYTES/SEC RD_BYTES/SEC WR_LAT   RD_LAT
rbd/ns/test 508/s 499/s    2.0 MiB/s    1.9 MiB/s 22.12 ms 4.63 ms
test1/test  487/s 475/s    1.9 MiB/s    1.9 MiB/s 23.46 ms 5.01 ms
```
I like more something like below:
```
Name        wr/s   rd/s    wr_bytes/s   rd_bytes/s   wr_lat   rd_lat
rbd/ns/test  508    499       2.0 MiB      1.9 MiB 22.12 ms  4.63 ms
test1/test   487    475       1.9 MiB      1.9 MiB 23.46 ms  5.01 ms
```
It looks more like system's `iostat`.

But please just ignore if you like your version more.

**Comment by dillaman** [comment]
> @dillaman If `iostat` or `iotop` started when there is no any rbd traffic the message "waiting for initial image stats" will be displayed until the traffic started. Why not display empty result, similarly to what we do when rbd traffic stopped when iostat/top is already running?

So that folks don't immediately assume it's broken since it can take up to 20 seconds for data to start coming in. But if you quit and restart it within a minute, data will appear immediately since the backend query hasn't timed out yet.

File: src/tools/rbd/action/Perf.cc [hunk 2]
```diff
@@ -0,0 +1,692 @@
...
+      }
+      tbl.define_column(title, TextTable::LEFT, TextTable::RIGHT);
+    }
+  }
+
+  for (auto& image_stat : image_stats) {
+    if (f)  {
+      f->open_object_section("image");
+      f->dump_string("pool", image_stat.pool_name);
+      f->dump_string("pool_space", image_stat.pool_namespace);
```
[the remark this answers is not in the document]
Comment (dillaman) [reply src/tools/rbd/action/Perf.cc]: Typo -- thx.

**Comment by trociny** [comment]
@dillaman 
```
qa/workunits/rbd/rbd-ggate.sh:    rbd namespace create ${POOL} ${NS}
qa/workunits/rbd/rbd-nbd.sh:    rbd namespace create ${POOL} ${NS}
```

File: src/pybind/mgr/rbd_support/module.py [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+                        if image_id not in image_names:
+                            self.log.debug("scrub_missing_images: dropping {}/{}".format(
+                                image_key, image_id))
+                            del sum_images[image_id]
+                            if image_id in raw_images:
+                                del raw_images[image_id]
+
+    def process_raw_osd_perf_counters(self):
+        now = datetime.now()
+        now_ts = int(now.timestamp())
```
Comment (trociny) [inline src/pybind/mgr/rbd_support/module.py]: for python2:
```
2019-01-28 13:09:36.265 7ff68dd60700  0 mgr[rbd_support] Fatal runtime error: 'datetime.datetime' object has no attribute 'timestamp'
Traceback (most recent call last):
  File "/home/mgolub/ceph/ceph.ci/src/pybind/mgr/rbd_support/module.py", line 137, in run
    self.process_raw_osd_perf_counters()
  File "/home/mgolub/ceph/ceph.ci/src/pybind/mgr/rbd_support/module.py", line 255, in process_raw_osd_perf_counters
    now_ts = int(now.timestamp())
AttributeError: 'datetime.datetime' object has no attribute 'timestamp'
```

Comment (trociny) [reply src/pybind/mgr/rbd_support/module.py]: `now_ts = int(now.strftime("%s"))` could be used.

Comment (dillaman) [reply src/pybind/mgr/rbd_support/module.py]: Thx -- updated

**Comment by AlexCompaq** [comment]
Hi,
How to use `rbd perf image iostat\iotop`? It always displayed the message "waiting for initial image stats".
Thx!

**Comment by dillaman** [comment]
It takes up to 30 seconds of IO activity in the pool before it will display.

**Comment by trociny** [comment]
Also, if rbd  is started when there is no rbd io activity in the cluster at all it may display the message longer until some activity starts.

**Pull request merged**